### ISSUE TITLE
portablectl inspect fails with "lacks os-release data" when the image attached is a directory

### ISSUE THREAD

**Issue by bluca** [systemd/systemd#14634]
**systemd version the issue has been seen with**

Both v239 and latest master (commit id fe888c79f50b)

**Used distribution**

Debian stable

**Expected behaviour you didn't see**

portablectl inspect successfully returns information about the image

**Unexpected behaviour you saw**

portablectl inspect fails with "lacks os-release data" error, despite os-release being there. Inspect on the exact same image wrapped in a squashfs and attached works fine.

**Steps to reproduce the problem**

Minimal repro steps:

```
$ mkdir -p squashfs/usr/lib/systemd/system squashfs/etc
$ cp /etc/os-release squashfs/usr/lib/
$ ln -sr squashfs/usr/lib/os-release squashfs/etc/
$ cat <<EOF>squashfs/usr/lib/systemd/system/test1.service
[Service]
Type=oneshot
ExecStart=echo hello one
EOF
$ cat <<EOF>squashfs/usr/lib/systemd/system/test2.service
[Service]
Type=oneshot
ExecStart=echo hello two
EOF
$ mksquashfs squashfs/ test.raw -all-root
$ sudo portablectl attach $PWD/squashfs/ test1.service
$ sudo portablectl attach $PWD/test.raw test2.service
```

Result:

```
$ portablectl list
NAME     TYPE      RO CRTIME MTIME                       USAGE STATE   
squashfs directory no n/a    n/a                         n/a   attached
test     raw       no n/a    Wed 2020-01-22 12:27:19 GMT 4.0K  attached

2 images listed.
$ portablectl inspect squashfs test1.service
(Matching unit files with prefixes 'test1.service'.)
Failed to inspect image metadata: Image '/etc/portables/squashfs' lacks os-release data, refusing.
$ portablectl inspect test test2.service
(Matching unit files with prefixes 'test2.service'.)
Image:
        /etc/portables/test.raw
Portable Service:
        n/a
Operating System:
        Debian GNU/Linux 10 (buster)
Unit files:
        test2.service
```

The error printed by portabled with debug level is: ```Couldn't acquire os-release file, ignoring: Not a directory```

Running it under gdb, it's the openat called in chase_symlinks in src/basic/fs-util.c that fails and returns "not a directory":

https://github.com/systemd/systemd/blob/fe888c79f50b6f40bb1918301daa2c4d87dcd7bf/src/basic/fs-util.c#L905

(with chase_symlinks being called by portabled via open_os_release: https://github.com/systemd/systemd/blob/fe888c79f50b6f40bb1918301daa2c4d87dcd7bf/src/shared/os-util.c#L39 )

**Comment by bluca** [comment]
```openat(fd, first + n, O_CLOEXEC|O_NOFOLLOW|O_PATH);``` fails precisely when:

```
Breakpoint 2, chase_symlinks (path=0x7ffd04cec720 "/etc/portables/squashfs/etc/os-release", 
    original_root=0x556e0b9cc5a0 "/etc/portables/squashfs", flags=1, ret_path=0x7ffd04cec9c0, ret_fd=0x7ffd04cec9b4)
    at ../src/basic/fs-util.c:908
908	                        if (errno == ENOENT &&
(gdb) p fd
$4 = 8
(gdb) p first
$5 = 0x556e0b9cdcd0 "/tmp"
(gdb) p n
$6 = 1
(gdb) p done
$7 = 0x556e0b9cfc20 "/etc/portables/squashfs"
(gdb) p todo
$8 = 0x556e0b9cddd4 "/squashfs/etc/os-release"
(gdb) p buffer
$9 = 0x556e0b9cddd0 "/tmp/squashfs/etc/os-release"
```

```
$ sudo readlink /proc/`systemctl show -p MainPID --value systemd-portabled`/fd/8
/etc/portables/squashfs
```

**Comment by bluca** [comment]
As far as I can see, this is what's happening:

- attaching a directory means symlinking it under /etc/portables/<name>, so in my case /etc/portables/squashfs is a symlink to the directory /tmp/squashfs
- chase_symlinks is called with the linked path, /etc/portables/squashfs/etc/os-release
- while chasing, it finds that /etc/portables/squashfs is a symlink to /tmp/squashfs
- given the symlink is absolute, to chase it it opens the root in fd (which will be used for openat()), but instead of "/" it uses "/etc/portables/squashfs" as the root in https://github.com/systemd/systemd/blob/fe888c79f50b6f40bb1918301daa2c4d87dcd7bf/src/basic/fs-util.c#L964
- it then tries to openat the first part ("tmp") of resolved path "/tmp/squashfs/etc/os-release" under  "/etc/portables/squashfs" which obviously doesn't work

**Comment by bluca** [comment]
By manually changing the symlink in /etc/portables to be relative rather than absolute, it then works as chase_symlinks does the right thing, which appears to confirm my suspicions above:

```
$ sudo rm -f /etc/portables/squashfs
$ sudo ln -sr /tmp/squashfs/ /etc/portables/
$ ls -l /etc/portables/
total 0
lrwxrwxrwx 1 root root 18 Jan 22 14:40 squashfs -> ../../tmp/squashfs
lrwxrwxrwx 1 root root 13 Jan 22 12:28 test.raw -> /tmp/test.raw
$ portablectl inspect squashfs test1.service
(Matching unit files with prefixes 'test1.service'.)
Image:
        /etc/portables/squashfs
Portable Service:
        n/a
Operating System:
        Debian GNU/Linux 10 (buster)
Unit files:
        test1.service
```

**Comment by bluca** [comment]
Resolving the symlink beforehand seems to work. I don't want to touch the logic in chase_symlinks, as I'm not familiar with it and it's used in a hundred other places and I can't test them all.

**Comment by poettering** [comment]
I prepped a fix in #14689. Any chance you could give that a whirl?

**Comment by bluca** [comment]
Tested on the system at $work where I found the issue, works like a charm. Thanks!
